crates/rpc-types-engine/src/superchain.rs

```diff
@@ -0,0 +1,180 @@
+use core::array::TryFromSliceError;
+
+use alloy_primitives::{B256, B64};
+
+/// Superchain Signal information.
+///
+/// The execution engine SHOULD warn the user when the recommended version is newer than the current
+/// version supported by the execution engine.
+///
+/// The execution engine SHOULD take safety precautions if it does not meet the required protocol
+/// version. This may include halting the engine, with consent of the execution engine operator.
+///
+/// See also: <https://specs.optimism.io/protocol/exec-engine.html#engine_signalsuperchainv1>
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
+pub struct SuperchainSignal {
+    /// The recommended Supercain Protocol Version.
+    pub recommended: ProtocolVersion,
+    /// The minimum Supercain Protocol Version required.
+    pub required: ProtocolVersion,
+}
+
+/// Formatted Superchain Protocol Version.
+///
+/// The Protocol Version documents the progression of the total set of canonical OP-Stack
+/// specifications. Components of the OP-Stack implement the subset of their respective protocol
+/// component domain, up to a given Protocol Version of the OP-Stack.
+///
+/// The Protocol Version **is NOT a hardfork identifier**, but rather indicates software-support for
+/// a well-defined set of features introduced in past and future hardforks, not the activation of
+/// said hardforks.
+///
+/// The Protocol Version is Semver-compatible. It is encoded as a single 32 bytes long
+/// protocol version. The version must be encoded as 32 bytes of DATA in JSON RPC usage.
+///
+/// See also: <https://specs.optimism.io/protocol/superchain-upgrades.html#protocol-version>
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ProtocolVersion {
+    /// Version-type 0.
+    V0(ProtocolVersionFormatV0),
+}
+
+#[derive(Copy, Clone, Debug, thiserror::Error)]
+pub enum ProtocolVersionError {
+    #[error("Unsupported version: {0}")]
+    UnsupportedVersion(u8),
+    #[error("Invalid version format length. Got {0}, expected 31")]
+    InvalidLength(usize),
+    #[error("Invalid version format encoding")]
+    FromSlice(#[from] TryFromSliceError),
+}
+
+impl From<ProtocolVersion> for B256 {
```

Comment:
please convert these to functions, these from conversions are horrible to work with.

Reply:
Got it, I thought the `[u8; 31]` was the issue